include/seqan3/range/views/take.hpp

```diff
@@ -283,22 +283,10 @@ class view_take : public std::ranges::view_interface<view_take<urng_t, exactly,
     /*!\name Associated types
      * \{
      */
-    //!\brief The reference_type.
-    using reference         = reference_t<urng_t>;
-    //!\brief The const_reference type is equal to the reference type if the underlying range is const-iterable.
-    using const_reference   = detail::transformation_trait_or_t<seqan3::reference<urng_t const>, void>;
-    //!\brief The value_type (which equals the reference_type with any references removed).
-    using value_type        = std::ranges::range_value_t<urng_t>;
-    //!\brief The size_type is `size_t` if the view is exact, otherwise `void`.
-    using size_type         = std::conditional_t<exactly || std::ranges::sized_range<urng_t>,
-                                                 transformation_trait_or_t<seqan3::size_type<urng_t>, size_t>,
-                                                 void>;
-    //!\brief A signed integer type, usually std::ptrdiff_t.
-    using difference_type   = std::ranges::range_difference_t<urng_t>;
     //!\brief The iterator type of this view (a random access iterator).
     using iterator          = iterator_type<urng_t>;
     //!\brief The const_iterator type is equal to the iterator type if the underlying range is const-iterable.
-    using const_iterator    = detail::transformation_trait_or_t<std::type_identity<iterator_type<urng_t const>>, void>;
+    using const_iterator    = iterator_type<urng_t const>;
```

Comment:
Both expressions do semantically the same:

```c++
iterator_type<urng_t const> // is a complete type, like int

std::type_identity<compete_type> 
```

is defined as:

https://github.com/seqan/seqan3/blob/c9b27e310fda52d8414efc0ff31af9f5df858b11/include/seqan3/std/type_traits#L36-L40

so it **always** has a `type` member.

```c++
transformation_trait_or_t<some_type, void>
```

returns `some_type::type` or void if `type` member is missing.

Since `std::type_identity` has always type it will always return the content of the argument.

---

I'm not sure if this has some wonky two-phase lookup semantics (a comment would be sometime golden if you use such a complicated construction).

ASFAIS there is nothing that really defers the template instantiation. Every single type (incomplete or not) are completely known at instantiation so that that expression will be evaluated as early as possible.

It seems to be an oversight, see the original expression:

https://github.com/seqan/seqan3/commit/402d35cc86bd847a5d5b229e806e87d90aea1c70#diff-fb9773f607908531fc1873edd3046821L264-R269

That expression was also quite simpler.

---

Normally, `std::type_identity` can be used for 

> type_identity can be used to establish non-deduced contexts in template argument deduction: 

https://en.cppreference.com/w/cpp/types/type_identity#Notes

Reply:
Thanks for the explanation. If all the tests pass I agree with the change since I don't see a reason for the need of std::type_identity. I'll ask Hannes for an explanation if he has time, if he doens't I'll just merge it.

Reply:
This is only a type def and a type def can declare stuff even if we don't use const_iterator internally. Let's keep this simple.

Reply:
I agree the current solution is not very clear, but I am pretty sure that it was necessary to defer instantiation of `iterator_type<urng_t const>` because `iterator_type` calls `iterator_t<X>` (should be `std::ranges::iterator_t`) on  `urng_t const` which leads to substitution failure if `!seqan3::const_iterable_range<urng_t>`.

Before removing this, please add a test-case for view take on a range that is not const-iterable.

> This is only a type def and a type def can declare stuff even if we don't use const_iterator internally. Let's keep this simple.

The member type is private for take  and public for take_until. If it stays private it doesn't make a difference, but it is public I would define the type only conditionally, because it is weird to have a `::const_iterator` that cannot be instantiated. But it's not that important.

Reply:
I think you already had test cases for this:

https://github.com/seqan/seqan3/blob/332f4b3749f7baeee1efb6e1e2cd310df0f2fdc0/test/unit/range/views/view_take_test.cpp#L127-L133

https://github.com/seqan/seqan3/blob/332f4b3749f7baeee1efb6e1e2cd310df0f2fdc0/test/unit/range/views/view_take_test.cpp#L235-L243

https://github.com/seqan/seqan3/blob/332f4b3749f7baeee1efb6e1e2cd310df0f2fdc0/test/unit/range/views/view_take_test.cpp#L248-L249

https://github.com/seqan/seqan3/blob/332f4b3749f7baeee1efb6e1e2cd310df0f2fdc0/test/unit/range/views/view_take_test.cpp#L290-L295

Reply:
There should also be a test that verifies that the type of `auto const & v2 = v` does not model the range concepts. It's important to ensure that this fails gracefully (concept-check == false) and does not lead to a build-failure. (Although I do think the implementation is now correct).